src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs

```diff
@@ -31,7 +31,8 @@ namespace MongoDB.Driver.Core.Authentication
     public abstract class SaslAuthenticator : IAuthenticator
     {
         // fields
-        private readonly ISaslMechanism _mechanism;
+        private protected readonly ISaslMechanism _mechanism;
+        private protected ISaslStep _speculativeFirstStep;
```

Comment:
Wouldn't `private protected` give us (the driver devs) the ability to subclass as needed while disallowing applications from accessing the field?

https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/private-protected

If we made the field simply `protected` then we'd essentially be exposing a new bit of public API, wouldn't we? E.g. Applications would be create a custom class that derives from `SaslAuthenticator` that could read and write from `_mechanism` and `_speculativeFirstStep`.

Reply:
Yes... private protected allows us (in *this* DLL only) to subclass this class and still be able to access these field in the subclasses.

My main point is that we are sending contradictory messages: this class is public with a protected constructor which signals that in theory at least an application could write their own subclass. But then we are sending a contradictory message by declaring the fields private protected.

Given that we don't actually want an application to subclass this class we can go with your changes to the field visibility.

Reply:
I think at least for _mechanism, there is no a lot of sense to make it `private internal`, we pass the mechanism in the protected constructor, so I don't see why it should be hidden. Also, maybe use properties?

Reply:
I think we typically only use properties when it's part of the public API? Since `_mechanism` and `_speculativeFirstStep` will NOT be exposed to the user, shouldn't using fields should be okay, similar to how we use fields instead of properties when they're private?

Unfortunately, while I would have loved to leave `_mechanism`  private, `ScramShaAuthenticator` needs access to it in the `CustomizeInitialIsMasterCommand` method. Theoretically we could potentially make `_speculativeFirstStep` private by moving it into `ScramShaAuthenticator` but then I think we'd need to refactor the SaslAuthenticator's Authenticate code to be virtual and add overriden logic to `ScramShaAuthenticator`, which I'd rather not to (since then we'd also be allowing the user to override the `Authenticate` logic)

Reply:
I think that it's not the best using of `_mechanism` field, but I can be ok with it, so resolved from my side